### PR TITLE
Implements Azure/vscode-aks-tools project #994:

### LINKED ISSUE #2104: Add `ContainerMode` enum and `MultiContainerWorkflowConfig` types
> **Note on Testing Plans:** All testing plans below are based on initial estimation and the design document. They should be treated as a starting point — please adapt, add, or remove test cases as you discover what's feasible in the current test infrastructure. If you hit blockers or need to discuss what's realistic given the existing test setup, ping in chat and we can figure out the right approach together.

## Description:

Add `ContainerMode` enum (`Single` / `Multi`) to `types.ts`. Add `ContainerJobConfig` and `MultiContainerWorkflowConfig` interfaces to `workflowTemplate.ts` to support per-service configuration (containerName, dockerFile, buildContextPath, deploymentManifestPath). Existing `WorkflowConfig` must remain unchanged (additive only).
- **Labels:** `enhancement`, `container-assist`
- **Acceptance Criteria:**
  - [ ] `ContainerMode.Single` and `ContainerMode.Multi` enum values exist
  - [ ] `ContainerJobConfig` interface with `containerName`, `dockerFile`, `buildContextPath`, `deploymentManifestPath`
  - [ ] `MultiContainerWorkflowConfig` interface with `containers: ContainerJobConfig[]` and shared Azure context fields
  - [ ] No breaking changes to existing `WorkflowConfig`
- **Testing Plan (estimated — adjust based on current test setup):**
  - [ ] **Unit:** Verify `ContainerMode.Single === "single"` and `ContainerMode.Multi === "multi"`
  - [ ] **Unit:** Instantiate a `ContainerJobConfig` object with all required fields; confirm TypeScript compilation passes
  - [ ] **Unit:** Instantiate a `MultiContainerWorkflowConfig` with an empty `containers` array and with multiple entries; confirm compilation
  - [ ] **Regression:** Import existing `WorkflowConfig` in a test file and verify it still compiles and works with all existing fields — no properties removed or renamed
  - [ ] **Compile-time:** Run `npm run compile` and confirm zero new errors

### LINKED ISSUE #2105: Show Container Mode QuickPick (Single / Multi) when ≥ 2 Dockerfiles detected
## Description

In `collectWorkflowConfiguration()` in `workflowGenerator.ts`, insert a radio-button-style QuickPick immediately before the Dockerfile selection step. The QuickPick should offer "Single Container (default)" and "Multi-Container (mono repo)" options. Only display this step when ≥ 2 Dockerfiles are detected in the workspace; otherwise skip and default to single-container flow.
- **Labels:** `enhancement`, `container-assist`, `UX`
- **Acceptance Criteria:**
  - [ ] QuickPick shown only when ≥ 2 Dockerfiles exist
  - [ ] "Single Container" is the default/pre-selected option
  - [ ] Selecting "Single Container" proceeds with existing single-container flow unchanged
  - [ ] Selecting "Multi-Container" proceeds to multi-select Dockerfile step
  - [ ] Canceling the QuickPick aborts the flow cleanly
- **Testing Plan (estimated — adjust based on current test setup):**
  - [ ] **Unit:** Mock `vscode.window.showQuickPick`; call with 0 Dockerfiles detected → assert QuickPick is **not** called, flow defaults to single mode
  - [ ] **Unit:** Mock with 1 Dockerfile detected → assert QuickPick is **not** called
  - [ ] **Unit:** Mock with 2 Dockerfiles detected → assert QuickPick **is** called with exactly 2 items ("Single Container", "Multi-Container")
  - [ ] **Unit:** Mock with 5 Dockerfiles detected → assert QuickPick is called
  - [ ] **Unit:** Simulate user selecting "Single Container" → assert returned mode is `ContainerMode.Single`
  - [ ] **Unit:** Simulate user selecting "Multi-Container" → assert returned mode is `ContainerMode.Multi`
  - [ ] **Unit:** Simulate user pressing Escape (QuickPick returns `undefined`) → assert function returns `undefined` / throws cancellation
  - [ ] **Integration (manual):** Open a mono-repo with 3 Dockerfiles → verify the prompt appears; select Single → verify original flow proceeds; repeat and select Multi → verify multi-select appears next

### LINKED ISSUE #2106: Replace single-select Dockerfile picker with multi-select when in Multi-Container mode
## Description:

When the user selects Multi-Container mode, show a `canPickMany: true` QuickPick listing all detected Dockerfiles. All Dockerfiles should be pre-checked by default. Each item's description should show language/framework/port metadata from the SDK module analysis when available. Validate that at least one Dockerfile is selected; show an error and re-prompt if the user unchecks all.
- **Labels:** `enhancement`, `container-assist`, `UX`
- **Acceptance Criteria:**
  - [ ] Multi-select QuickPick with `canPickMany: true`
  - [ ] All detected Dockerfiles pre-checked
  - [ ] Description shows SDK metadata (language, port) when available
  - [ ] Validation error if zero Dockerfiles selected
  - [ ] Cancel aborts the flow cleanly
- **Testing Plan (estimated — adjust based on current test setup):**
  - [ ] **Unit:** Mock `showQuickPick` with `canPickMany: true`; pass 3 Dockerfiles → assert all 3 items have `picked: true`
  - [ ] **Unit:** Verify QuickPick item labels are relative Dockerfile paths (e.g., `frontend/Dockerfile`)
  - [ ] **Unit:** Verify QuickPick item descriptions contain SDK metadata when present (e.g., `"React, port 3000"`)
  - [ ] **Unit:** Verify descriptions are empty/undefined when SDK metadata is unavailable
  - [ ] **Unit:** Simulate user selecting 2 of 3 Dockerfiles → assert returned array has exactly 2 entries
  - [ ] **Unit:** Simulate user unchecking all and confirming → assert validation error message is shown and QuickPick is re-displayed
  - [ ] **Unit:** Simulate user pressing Escape → assert flow cancels cleanly
  - [ ] **Regression:** In single-container mode with ≥ 2 Dockerfiles, assert the existing single-select QuickPick (`canPickMany: false`) is still used
  - [ ] **Integration (manual):** Open mono-repo, select Multi-Container mode → visually verify all Dockerfiles are checked, metadata shows, uncheck all → verify error re-prompts

### LINKED ISSUE #2107: Auto-derive containerName, buildContextPath, and manifest paths per selected Dockerfile
## Description

For each Dockerfile selected in multi-container mode, automatically derive: `containerName` from the SDK module name (e.g., `frontend`), `buildContextPath` as the parent directory of the Dockerfile (e.g., `frontend/`), and `manifestPaths` by scanning `<moduleDir>/k8s/` for YAML files. No additional user prompts should be required when all values can be auto-detected.
- **Labels:** `enhancement`, `container-assist`
- **Acceptance Criteria:**
  - [ ] `containerName` derived from module/directory name
  - [ ] `buildContextPath` set to parent directory of Dockerfile
  - [ ] `manifestPaths` auto-detected from `<moduleDir>/k8s/` directory
  - [ ] No extra prompts when all values are auto-detected
  - [ ] Works correctly with `hasBothActions` flow (deployment + workflow)
- **Testing Plan (estimated — adjust based on current test setup):**
  - [ ] **Unit:** Given Dockerfile at `frontend/Dockerfile` → assert `containerName === "frontend"` and `buildContextPath === "frontend"`
  - [ ] **Unit:** Given Dockerfile at `services/api/Dockerfile` → assert `containerName === "api"` and `buildContextPath === "services/api"`
  - [ ] **Unit:** Given Dockerfile at root `./Dockerfile` → assert `containerName` falls back to project/repo name and `buildContextPath === "."`
  - [ ] **Unit:** Mock filesystem with `frontend/k8s/deployment.yaml` and `frontend/k8s/service.yaml` → assert both paths returned in `manifestPaths`
  - [ ] **Unit:** Mock filesystem with no `<module>/k8s/` directory → assert `manifestPaths` is empty (triggers Browse prompt in Work Item 5)
  - [ ] **Unit:** Mock `hasBothActions === true` with deployment output containing 3 modules → assert all per-service config auto-derived with zero `showQuickPick` / `showInputBox` calls
  - [ ] **Unit:** Verify `containerName` is sanitized for use in YAML job names (no spaces, special chars)
  - [ ] **Integration (manual):** Generate deployment files for a 3-service mono-repo, then generate workflow → verify all config auto-filled, no extra prompts shown

### LINKED ISSUE #2108: Add Browse / Enter Manually / Skip dialog for missing K8s manifests
## Description

Replace the free-text `InputBox` prompt for missing K8s manifests with a three-button information message dialog: "Browse..." (opens `showOpenDialog` with YAML filters rooted at workspace), "Enter path manually" (falls back to existing InputBox), and "Skip" (omits manifests for that service). This improvement applies to both single-container and multi-container modes. Add a `promptForManifestWithBrowse()` helper function.
- **Labels:** `enhancement`, `container-assist`, `UX`
- **Acceptance Criteria:**
  - [ ] Three-button dialog: Browse, Enter path manually, Skip
  - [ ] Browse opens `showOpenDialog` with `canSelectMany: true` and YAML file filters
  - [ ] Selected paths are correctly converted to workspace-relative paths
  - [ ] "Enter path manually" falls back to existing InputBox
  - [ ] "Skip" returns empty array, omitting manifests from workflow
  - [ ] Works in both single-container and multi-container modes
- **Testing Plan (estimated — adjust based on current test setup):**
  - [ ] **Unit:** Mock `showInformationMessage` → assert it is called with three button labels: "Browse...", "Enter path manually", "Skip"
  - [ ] **Unit:** Mock `showInformationMessage` returning "Browse..." → assert `showOpenDialog` is called with `canSelectMany: true` and filters `{ "Kubernetes Manifests": ["yaml", "yml"] }`
  - [ ] **Unit:** Mock `showOpenDialog` returning 2 URIs → assert returned paths are workspace-relative (e.g., `api/k8s/deploy.yaml` not absolute)
  - [ ] **Unit:** Mock `showOpenDialog` returning `undefined` (user cancelled dialog) → assert function returns `undefined`
  - [ ] **Unit:** Mock `showInformationMessage` returning "Enter path manually" → assert `showInputBox` is called (existing fallback)
  - [ ] **Unit:** Mock `showInformationMessage` returning "Skip" → assert function returns empty array `[]`
  - [ ] **Unit:** Mock `showInformationMessage` returning `undefined` (user dismissed) → assert function returns `undefined`
  - [ ] **Unit:** Verify service name is displayed in the dialog message (e.g., `"No K8s manifests found for "api""`)
  - [ ] **Regression:** In single-container mode with no manifests detected → verify the same Browse/Manual/Skip dialog appears (replaces old InputBox)
  - [ ] **Integration (manual):** Delete `k8s/` directory from one service, trigger workflow generation → verify dialog appears for that service, test each button

### LINKED ISSUE #2109: Add helper to scan manifests per module path in `fileOperations.ts`
## Description

Add a function in `fileOperations.ts` that, given a module directory path, scans `<moduleDir>/k8s/` for `.yaml` and `.yml` files and returns their paths. This helper is used by the per-service configuration loop in multi-container mode to auto-detect manifests for each selected Dockerfile's module.
- **Labels:** `enhancement`, `container-assist`
- **Acceptance Criteria:**
  - [ ] Function accepts a module directory path and workspace root
  - [ ] Scans `<moduleDir>/k8s/` for `.yaml` and `.yml` files
  - [ ] Returns array of relative paths (empty array if directory doesn't exist)
  - [ ] Handles edge cases: no k8s directory, empty k8s directory, nested manifests
- **Testing Plan (estimated — adjust based on current test setup):**
  - [ ] **Unit:** Create temp directory with `frontend/k8s/deployment.yaml` and `frontend/k8s/service.yml` → assert both files returned
  - [ ] **Unit:** Create temp directory with empty `frontend/k8s/` → assert empty array returned
  - [ ] **Unit:** Create temp directory with no `frontend/k8s/` directory → assert empty array returned (no error thrown)
  - [ ] **Unit:** Create temp directory with `frontend/k8s/subfolder/nested.yaml` → assert nested file is **not** returned (flat scan only) or **is** returned (document which behavior is expected)
  - [ ] **Unit:** Create temp directory with `frontend/k8s/readme.md` and `frontend/k8s/deploy.yaml` → assert only `.yaml`/`.yml` files returned
  - [ ] **Unit:** Verify returned paths are relative to workspace root
  - [ ] **Unit:** Verify function handles symlinks gracefully (no infinite loops)
  - [ ] **Edge case:** Module directory path contains spaces (e.g., `my service/k8s/`) → assert correct paths returned

### LINKED ISSUE #2110: Add multi-job GitHub Actions workflow template for multi-container deployments
## Description

Create `resources/yaml/aks-deploy-multi.template.yaml` (or implement dynamic generation in `workflowTemplate.ts`) that renders a combined workflow with one build-push-deploy job per selected container. Each job should reference the correct Dockerfile, build context, ACR image name, and K8s manifest paths. The template must produce valid GitHub Actions syntax.
- **Labels:** `enhancement`, `container-assist`
- **Acceptance Criteria:**
  - [ ] Template generates one job per container (e.g., `build-frontend`, `build-api`)
  - [ ] Each job references correct Dockerfile path and build context
  - [ ] Each job pushes to ACR with a service-specific image name
  - [ ] Each job deploys the correct K8s manifests
  - [ ] Shared Azure context (ACR, cluster, namespace) used across jobs
  - [ ] Generated YAML is valid GitHub Actions syntax
- **Testing Plan (estimated — adjust based on current test setup):**
  - [ ] **Unit:** Render template with 1 container config → assert 1 job block generated with correct name, Dockerfile, and context
  - [ ] **Unit:** Render template with 3 container configs → assert 3 job blocks generated, each with unique job ID (e.g., `build-frontend`, `build-api`, `build-worker`)
  - [ ] **Unit:** Verify each job's `docker build` step references the correct Dockerfile path (e.g., `frontend/Dockerfile`)
  - [ ] **Unit:** Verify each job's `docker build` step uses the correct build context (e.g., `frontend/`)
  - [ ] **Unit:** Verify each job's `docker push` step pushes to `<acr>.azurecr.io/<containerName>:${{ github.sha }}`
  - [ ] **Unit:** Verify each job's deploy step references the correct manifest paths
  - [ ] **Unit:** Verify shared fields (`on.push.branches`, environment variables for ACR/cluster/namespace) are correct
  - [ ] **Validation:** Parse generated YAML with a YAML parser library → assert no syntax errors
  - [ ] **Validation:** Validate generated YAML against GitHub Actions schema (or at minimum check for required keys: `name`, `on`, `jobs`, `jobs.<id>.runs-on`, `jobs.<id>.steps`)
  - [ ] **Snapshot:** Add a snapshot test comparing rendered output for a 2-container config against a golden file
  - [ ] **Edge case:** Container name with hyphens/underscores → verify valid YAML job ID generated

### LINKED ISSUE #2111: Add multi-container workflow rendering logic in `workflowTemplate.ts`
## Description

Extend `workflowTemplate.ts` to accept `MultiContainerWorkflowConfig` and render the multi-job workflow YAML. The rendering should iterate over `config.containers` and generate a separate job block for each, substituting per-service values (containerName, dockerFile, buildContextPath, manifestPaths) into the template. The existing single-container rendering must remain unchanged.
- **Labels:** `enhancement`, `container-assist`
- **Acceptance Criteria:**
  - [ ] New rendering function accepts `MultiContainerWorkflowConfig`
  - [ ] Generates one job per container in the `containers` array
  - [ ] Per-job values correctly substituted (image name, Dockerfile, context, manifests)
  - [ ] Existing single-container rendering unchanged
  - [ ] Output written to `.github/workflows/<name>.yml`
- **Testing Plan (estimated — adjust based on current test setup):**
  - [ ] **Unit:** Call new render function with a `MultiContainerWorkflowConfig` containing 2 containers → assert output contains 2 distinct job blocks
  - [ ] **Unit:** Verify placeholder substitution: `{{DOCKER_FILE}}` → `frontend/Dockerfile`, `{{BUILD_CONTEXT}}` → `frontend`, etc. for each job
  - [ ] **Unit:** Verify the `workflowName` is used in the YAML `name:` field
  - [ ] **Unit:** Verify `branchName` appears in `on.push.branches`
  - [ ] **Unit:** Verify ACR, cluster, resource group, namespace values appear in environment variables or step parameters
  - [ ] **Unit:** Verify `isManagedNamespace` flag correctly toggles namespace creation step
  - [ ] **Regression:** Call the existing single-container render function with a `WorkflowConfig` → assert output is **identical** to previous behavior (no regressions)
  - [ ] **Regression:** Run all existing workflow template tests → assert they pass without modification
  - [ ] **Integration:** Generate a multi-container workflow, write to disk, read back and parse YAML → assert structure matches expected jobs
  - [ ] **Edge case:** Render with a single container in `containers[]` → assert valid single-job workflow produced

### LINKED ISSUE #2112: Handle multi-container edge cases (cancel, single-check, overwrite, hasBothActions)
## Description

Implement edge case handling as defined in the design document: (1) If user unchecks all Dockerfiles, show validation error and re-prompt. (2) If user selects Multi but checks only 1 Dockerfile, generate a single-job workflow. (3) Existing workflow file overwrite confirmation applies unchanged. (4) When `hasBothActions` is true, auto-derive all per-service config from deployment output without extra prompts. (5) Clean exit when user cancels at any step (mode selection, Dockerfile selection, manifest browse).
- **Labels:** `enhancement`, `container-assist`, `bug-prevention`
- **Acceptance Criteria:**
  - [ ] Zero-selection validation with re-prompt
  - [ ] Single Dockerfile in multi-mode produces single-job workflow
  - [ ] Overwrite confirmation works for multi-container workflows
  - [ ] `hasBothActions` auto-derives config, skips prompts
  - [ ] Cancel at any step exits cleanly without partial state
- **Testing Plan (estimated — adjust based on current test setup):**
  - [ ] **Unit — Zero selection:** Mock `showQuickPick` returning empty array → assert error message shown via `showWarningMessage` and `showQuickPick` called a second time
  - [ ] **Unit — Single in multi:** Mock user selecting only 1 Dockerfile in multi mode → assert `MultiContainerWorkflowConfig.containers` has 1 entry → assert rendered workflow has exactly 1 job
  - [ ] **Unit — Overwrite:** Mock workflow file already exists at target path → assert `showWarningMessage` with overwrite confirmation is called → if user confirms, file is overwritten; if user declines, flow aborts
  - [ ] **Unit — hasBothActions:** Mock `hasBothActions === true` with deployment output containing `[{module: "frontend", ...}, {module: "api", ...}]` → assert zero `showQuickPick` / `showInputBox` / `showInformationMessage` calls for Dockerfile/manifest selection
  - [ ] **Unit — Cancel at mode selection:** Mock `showQuickPick` for mode returning `undefined` → assert no files written, no side effects
  - [ ] **Unit — Cancel at Dockerfile selection:** Mock mode = Multi, then `showQuickPick` for Dockerfiles returning `undefined` → assert clean exit
  - [ ] **Unit — Cancel at manifest browse:** Mock Browse dialog returning `undefined` → assert clean exit
  - [ ] **Unit — No partial state:** After any cancellation, assert no workflows files created or modified
  - [ ] **Integration (manual):** Walk through the full flow canceling at each step; verify extension returns to idle state without errors in the Output panel

### LINKED ISSUE #2113: Add test coverage for multi-container workflow generation flow
## Description

Add tests covering the multi-container flow: (1) Single-container regression — existing flow unchanged with 0–1 Dockerfiles. (2) Mode prompt visibility — radio appears only when ≥ 2 Dockerfiles. (3) Multi-select pre-check behavior. (4) Per-service manifest auto-detection from `<module>/k8s/`. (5) Browse dialog returns valid paths. (6) Generated multi-job YAML is valid GitHub Actions syntax with correct per-job values. (7) Cancel at any step produces clean exit. (8) `hasBothActions` auto-derive path.
- **Labels:** `testing`, `container-assist`
- **Acceptance Criteria:**
  - [ ] Single-container regression test passes
  - [ ] Mode prompt visibility test (shown at ≥ 2, hidden at 0–1)
  - [ ] Multi-select pre-check and uncheck-all validation tests
  - [ ] Per-service manifest auto-detect test
  - [ ] Browse dialog integration test
  - [ ] Generated YAML validation test (valid syntax, correct refs)
  - [ ] Cancel/abort tests at each step
  - [ ] `hasBothActions` auto-derive test
- **Testing Plan (estimated — adjust based on current test setup):**
  - [ ] **Test file:** Create `src/tests/suite/containerAssist/multiContainerWorkflow.test.ts`
  - [ ] **Test: Single-container regression** — Mock 1 Dockerfile detected → run `collectWorkflowConfiguration()` → assert no mode QuickPick shown, existing single-select Dockerfile QuickPick used, generated workflow identical to current behavior
  - [ ] **Test: Zero Dockerfiles regression** — Mock 0 Dockerfiles → assert flow proceeds to free-text InputBox (existing behavior), no mode prompt
  - [ ] **Test: Mode prompt shown** — Mock 2 Dockerfiles → assert `showQuickPick` called with mode options
  - [ ] **Test: Mode prompt shown (many)** — Mock 10 Dockerfiles → assert same behavior as 2
  - [ ] **Test: Multi-select pre-checked** — Assert all QuickPick items have `picked: true`
  - [ ] **Test: Uncheck-all validation** — Mock empty selection → assert re-prompt loop
  - [ ] **Test: Auto-detect manifests** — Set up temp workspace with `frontend/k8s/deploy.yaml`, `api/k8s/service.yaml` → assert correct paths returned per service
  - [ ] **Test: Browse dialog — Browse path** — Mock Browse → showOpenDialog returning URIs → assert relative paths
  - [ ] **Test: Browse dialog — Manual path** — Mock "Enter path manually" → assert InputBox called
  - [ ] **Test: Browse dialog — Skip** — Mock "Skip" → assert empty array
  - [ ] **Test: YAML validity** — Render 3-container template → parse with YAML library → assert no parse errors, assert `jobs` has 3 keys
  - [ ] **Test: YAML correctness** — Assert each job references its specific Dockerfile, build context, image name, manifest path
  - [ ] **Test: Cancel at mode** — Mock mode QuickPick → `undefined` → assert clean exit
  - [ ] **Test: Cancel at Dockerfile select** — Mock Dockerfile QuickPick → `undefined` → assert clean exit
  - [ ] **Test: Cancel at browse** — Mock browse → `undefined` → assert clean exit
  - [ ] **Test: hasBothActions** — Mock deployment output with 3 modules → assert config auto-derived for all 3 with zero user prompts
  - [ ] **Test: Job ID sanitization** — Container name `my-service_v2` → assert valid YAML key in job ID
  - [ ] **Coverage target:** ≥ 80% line coverage for new code in `workflowGenerator.ts`, `workflowTemplate.ts`, `fileOperations.ts`

---

> **Reminder:** These testing plans are based on estimation from the design document. The actual test cases, mocking strategies, and feasibility may differ depending on the current test infrastructure (test runner, mock utilities, CI setup, etc.). Please adapt as you go — add or remove tests based on what's practical. If anything looks unclear or you're unsure what's doable in the current setup, ping in chat and we'll work through it together.

### PR DESCRIPTION
**Pull request by Tatsinnit**
## Multi-Container (mono-repo) GitHub Workflow generation

Implements Azure project [#994](https://github.com/orgs/Azure/projects/994) — issues #2104 – #2113.

Container Assist's "Generate GitHub Workflow" flow now supports mono-repos with multiple services. When 2+ Dockerfiles are detected, the user can opt into **Multi-Container mode**, multi-select which Dockerfiles to include, and get one workflow file with a `build-<service>` + `deploy-<service>` job pair per container.

### Motivation

Previously, even when Container Assist detected multiple modules and generated multiple Dockerfiles + manifests, the workflow step asked the user to pick **one** Dockerfile and produced **one** build/deploy pair. Developers had to manually duplicate and edit the generated YAML — error-prone and a poor experience for mono-repos like `frontend/ + api/ + worker/`.

#### UX flow

```
Detect Dockerfiles
   │
   ├─ 0–1 found → existing single-container flow (unchanged)
   │
   └─ ≥ 2 found → Container Mode QuickPick
                    ├─ Single → existing single-container flow
                    └─ Multi  → multi-select Dockerfiles (all pre-checked,
                                enriched with language / framework / port)
                                  │
                                  ├─ Per-service auto-derive
                                  │     containerName, buildContextPath, dockerFile
                                  │
                                  ├─ Per-service manifest scan in <module>/k8s/
                                  │     └─ If missing → Browse… / Enter manually / Skip
                                  │
                                  └─ Render multi-job workflow → .github/workflows/<name>.yml
```

#### Generated YAML shape

- Workflow-level `env:` for shared values — `ACR_RESOURCE_GROUP`, `AZURE_CONTAINER_REGISTRY`, `CLUSTER_NAME`, `CLUSTER_RESOURCE_GROUP`, `NAMESPACE`.
- Per-job `env:` for per-service values — `CONTAINER_NAME`, `DOCKER_FILE`, `BUILD_CONTEXT_PATH`, `DEPLOYMENT_MANIFEST_PATH`.
- `deploy-<service>` `needs: [build-<service>]` so services build & deploy independently inside the same workflow.
- Managed-namespace branch reuses the existing `az aks namespace get-credentials` + `kubelogin` flow from the single-container template.

### Files changed

| File | Purpose |
|---|---|
| workflowTemplate.ts | New `ContainerMode` enum, `ContainerJobConfig`, `MultiContainerWorkflowConfig`, `renderMultiContainerWorkflowTemplate`, `validateMultiContainerWorkflowConfig`, plus a `sanitizeJobId` helper. |
| workflowGenerator.ts | Mode dispatcher in `collectWorkflowConfiguration`; new `promptForContainerMode`, `promptForMultiDockerfileSelection`, `promptForManifestWithBrowse`, `collectMultiContainerWorkflowConfiguration`. Existing `promptForManifestSelection` now delegates to the Browse dialog when no manifests are detected (used by both modes). |
| fileOperations.ts | New `scanManifestsForModulePaths(modulePaths, k8sFolder)` returning `Map<modulePath, manifestPaths>` for parallel per-module scanning. |
| workflowTemplate.test.ts | 13 new tests covering multi-container rendering + validation. |

### Issue coverage

| Issue | Status |
|---|---|
| #2104 — Add `ContainerMode` enum and `MultiContainerWorkflowConfig` types | ✅ |
| #2105 — Show Container Mode QuickPick when ≥ 2 Dockerfiles detected | ✅ |
| #2106 — Replace single-select Dockerfile picker with multi-select in Multi mode | ✅ |
| #2107 — Auto-derive `containerName`, `buildContextPath`, manifest paths per Dockerfile | ✅ |
| #2108 — Add Browse / Enter Manually / Skip dialog for missing K8s manifests | ✅ |
| #2109 — Add helper to scan manifests per module path in fileOperations.ts | ✅ |
| #2110 — Multi-job GitHub Actions workflow template for multi-container deployments | ✅ |
| #2111 — Multi-container workflow rendering logic in workflowTemplate.ts | ✅ |
| #2112 — Edge cases (cancel, single-check re-prompt, overwrite, `hasBothActions`) | ✅ |
| #2113 — Test coverage for multi-container workflow generation flow | ✅ |

### Edge cases handled

- **0–1 Dockerfiles** → existing single-container flow runs unchanged (no mode prompt).
- **All Dockerfiles unchecked** in multi-select → warning + re-prompt once; second empty selection cancels cleanly.
- **Missing manifests for a service** → Browse / Enter manually / Skip dialog.
- **Skip** in Multi mode → emits a build-only job (`deploy-<service>` is omitted).
- **Duplicate container names** (e.g. two services sharing basename) → de-duplicated with `-2`, `-3` suffixes.
- **Non-identifier container names** (`My Service@v1`) → sanitized to valid GitHub Actions job ids matching `^[A-Za-z_][A-Za-z0-9_-]*$`.
- **Workflow file already exists** → existing overwrite confirmation applies unchanged.
- **`hasBothActions`** (deployment + workflow chosen together) → Single-mode auto-derivation is preserved exactly.

### Backward compatibility

- No changes to `WorkflowConfig` or single-container rendering — verified by all pre-existing tests passing.
- `MultiContainerWorkflowConfig` is additive; no template files were modified.
- The Browse / Manual / Skip dialog now also benefits single-container users when their project has no detectable manifests (previously: a plain free-text input box).

### Tests

```
182 passing (1s)
```

New `Multi-Container Workflow Template Tests` describe block covers:

- Per-container build + deploy job pair generation
- Workflow-level vs. per-job env placement
- No unreplaced `{ { X } }` placeholders
- Preserved GitHub Actions `${{ }}` expressions
- Managed vs. non-managed namespace branches
- Skip → deploy job omitted, build job retained
- Container name → sanitized job id
- Validation: empty containers, missing per-container fields, duplicate job ids, missing workflow-level fields

### Out of scope (deferred)

- Per-service workflow files with `paths:` triggers (`deploy-frontend.yml`, `deploy-api.yml`).
- Inter-service `needs:` ordering (e.g. DB migrations → API).
- Docker Compose detection.
- Webview-based multi-container configuration wizard.

These are listed as future work in MULTI_CONTAINER_DESIGN.md §10.

### How to test

1. Open a mono-repo with ≥ 2 Dockerfiles (e.g. `frontend/Dockerfile`, `api/Dockerfile`).
2. Run **AKS: Container Assist → Deploy to AKS**.
3. Pick **Generate GitHub Workflow**.
4. Confirm the **Container Mode** QuickPick appears; pick **Multi-Container**.
5. Confirm the Dockerfile multi-select shows all detected files with language/framework/port descriptions, all pre-checked.
6. For a service with no `<module>/k8s/` manifests, confirm the **Browse… / Enter manually / Skip** dialog appears.
7. Open the generated `.github/workflows/<name>.yml` and verify it contains `build-<service>` + `deploy-<service>` pairs with correct per-job env values.

### CODE REVIEW

File: src/commands/aksContainerAssist/workflowTemplate.ts [hunk 1]
```diff
@@ -144,3 +184,210 @@ export function validateWorkflowConfig(config: WorkflowConfig): string[] {
...
+    }
+
+    return errors;
+}
+
+/**
+ * Build the per-container build job YAML fragment.
+ * Indented with 4 spaces (matching the existing single-container template style).
+ */
+function renderBuildJob(jobId: string, container: ContainerJobConfig): string {
```
Comment (bosesuneha) [inline src/commands/aksContainerAssist/workflowTemplate.ts:283]: Can we use templates similar to single container workflow for the yaml code?

File: src/commands/aksContainerAssist/workflowTemplate.ts [hunk 2]
```diff
@@ -144,3 +184,210 @@ export function validateWorkflowConfig(config: WorkflowConfig): string[] {
...
+    ACR_RESOURCE_GROUP: ${config.acrResourceGroup}
+    AZURE_CONTAINER_REGISTRY: ${config.azureContainerRegistry}
+    CLUSTER_NAME: ${config.clusterName}
+    CLUSTER_RESOURCE_GROUP: ${config.clusterResourceGroup}
+    NAMESPACE: ${config.namespace}
+
+jobs:
+`;
+
+    const jobs = config.containers
```
Comment (bosesuneha) [inline src/commands/aksContainerAssist/workflowTemplate.ts:446]: we can use matrix strategy instead for the build jobs

File: src/commands/aksContainerAssist/workflowGenerator.ts [hunk 3]
```diff
@@ -178,24 +222,104 @@ async function collectWorkflowConfiguration(
...
+    const manifestsByModule = await scanManifestsForModulePaths(moduleScanRoots);
+
+    const containers: ContainerJobConfig[] = [];
+    for (const item of selected) {
+        const dockerfileAbsolute = path.resolve(projectRoot, item.dockerfileRelToProject);
+        const buildContextAbsolute = path.dirname(dockerfileAbsolute);
+        const buildContextRelToWorkspace = toPosixPath(path.relative(workspaceRoot, buildContextAbsolute)) || ".";
+        const dockerfileRelToBuildContext = toPosixPath(path.relative(buildContextAbsolute, dockerfileAbsolute));
+
+        const moduleManifests = manifestsByModule.get(buildContextAbsolute) ?? [];
```
Comment (bosesuneha) [inline src/commands/aksContainerAssist/workflowGenerator.ts:290]: some repos can have a single manifest for all services. are we going to support that?

File: src/commands/aksContainerAssist/workflowTemplate.ts (hunk 1 above)
Comment (Tatsinnit) [reply src/commands/aksContainerAssist/workflowTemplate.ts:283]: Great point. The single-container path uses `resources/yaml/aks-deploy.template.yaml` with placeholder substitution, and ideally the multi-container path would too. I kept it as TS template strings in this PR because the multi-container workflow needs a variable number of `build-*` / `deploy-*` jobs (one pair per service), each with its own `env:` block and conditional managed-namespace step — that's awkward to express as a single static YAML template with simple `{{var}}` substitution.

I'd like to address this as a follow-up: factor out per-job fragments (build job, deploy job, managed-NS annotate step) into template files under `resources/yaml/` and have the renderer concatenate them. Happy to open an issue to track it, does that work?

File: src/commands/aksContainerAssist/workflowTemplate.ts (hunk 2 above)
Comment (Tatsinnit) [reply src/commands/aksContainerAssist/workflowTemplate.ts:446]: Agreed that `strategy.matrix` would make the build side much cleaner. Two reasons I didn't do it here:

1. The deploy jobs can't easily share a matrix — each service can have a different `DEPLOYMENT_MANIFEST_PATH` (potentially a multi-line block scalar) and the managed-namespace branch adds an extra step for some services but not others, so deploy ends up needing per-service jobs anyway.
2. Build needs per-service `DOCKER_FILE` / `BUILD_CONTEXT_PATH` / `CONTAINER_NAME`, which fits a matrix, but mixing a matrix build + N discrete deploys (with `needs: build-<job>`) makes the graph harder to read than the current symmetric `build-<svc>` → `deploy-<svc>` pairs.

I think it's worth revisiting once we settle the template question above, happy to open a follow-up issue. Let me know if you'd like me to switch builds to matrix in this PR regardless.

File: src/commands/aksContainerAssist/workflowGenerator.ts (hunk 3 above)
Comment (Tatsinnit) [reply src/commands/aksContainerAssist/workflowGenerator.ts:290]: Good call-out. Today the flow asks for a manifest path per service, so a shared manifest works but the user has to re-select the same file N times — functional but clunky.

Proposed follow-up UX: after the first service's manifest is selected, offer a quick pick — "Apply this manifest to all remaining services" / "Choose individually". That keeps the per-service model (in case some services need to opt out) while supporting the common single-manifest case. Want me to file this as a follow-up issue, or would you prefer it land in this PR?

File: src/commands/aksContainerAssist/workflowTemplate.ts (hunk 1 above)
Comment (Tatsinnit) [reply src/commands/aksContainerAssist/workflowTemplate.ts:283]: Filed as #2161 to track.

File: src/commands/aksContainerAssist/workflowTemplate.ts (hunk 2 above)
Comment (Tatsinnit) [reply src/commands/aksContainerAssist/workflowTemplate.ts:446]: Filed as #2162 to track.

File: src/commands/aksContainerAssist/workflowGenerator.ts (hunk 3 above)
Comment (Tatsinnit) [reply src/commands/aksContainerAssist/workflowGenerator.ts:290]: Filed as #2163 to track.

**Review by bosesuneha — APPROVED** [review]
approving this, we can track the changes in the new issues
